File: ask.py
"""Ask mode: a model works the history tools in-process and synthesizes
an answer (wip/SPEC-ask-mode.md). Provider-agnostic behind two adapters —
"openai-compatible" (OpenAI, OpenRouter, Groq, Ollama's /v1, …) and
"anthropic" — selected per named preset in [ask.models]; the client only
ever picks a preset name, so no query param can point the loop (and a
key) at an arbitrary endpoint.

Raw HTTP via requests, no provider SDKs — the embed-backend pattern.
Tools are read-only, so injected instructions in indexed content can at
worst produce a bad answer.
"""
import json, os, re
from datetime import datetime

import requests

import config
import server

TOOL_RESULT_MAX = 20_000     # chars per tool result fed back to the model
PROVIDER_TIMEOUT = 90        # seconds per provider call

CITE_RE = re.compile(r"\[id:([^\]\s]+)\]")


class AskError(RuntimeError):
    pass


def presets() -> list:
    """[ask.models] entries whose key is actually present (an empty
    key_env means keyless — always available). On the Lambda there is no
    TOML, so the list may arrive as JSON in CLAUDE_RAG_ASK_MODELS."""
    raw = config.get("ask", "models", "CLAUDE_RAG_ASK_MODELS", [])
    if isinstance(raw, str):
        try:
            raw = json.loads(raw)
        except json.JSONDecodeError:
            return []
    out = []
    for m in raw or []:
        if not m.get("name") or not m.get("model"):
            continue
        key_env = m.get("key_env", "")
        key = os.environ.get(key_env, "") if key_env else ""
        if key_env and not key:
            continue
        out.append({**m, "_key": key})
    return out


def tool_defs() -> list:
    """Neutral tool definitions derived from the MCP registry — the
    docstrings already teach the disclosure ladder, and derived schemas
    can't drift from the tools."""
    return [{"name": t.name, "description": t.description,
             "schema": t.parameters}
            for t in server.mcp._tool_manager.list_tools()]


def run_tool(name: str, args: dict) -> str:
    fn = getattr(server, name, None)
    if fn is None or not any(t["name"] == name for t in tool_defs()):
        return json.dumps({"error": f"unknown tool {name!r}"})
    try:
        result = fn(**(args or {}))
    except Exception as e:
        return json.dumps({"error": f"{name} failed: {e}"})
    if len(result) > TOOL_RESULT_MAX:
        result = result[:TOOL_RESULT_MAX] + "… [truncated]"
    return result


def _system_prompt() -> str:
    now = datetime.now().astimezone()
    return (
        "You answer questions from the user's own indexed history — their "
        "Claude Code sessions, shell commands, browsing, git commits, "
        "notes, calendar, and app usage — using the provided tools.\n"
        f"Now: {now.isoformat(timespec='minutes')} "
        f"({now.tzname()}). Resolve relative dates against that.\n"
        "Work the disclosure ladder: history_stats to orient when unsure "
        "what's indexed; search_history / list_window to find (read digest "
        "chunks first for day/week questions); expand to read a hit in "
        "full before leaning on it.\n"
        "Cite the chunks your answer rests on inline as [id:<chunk id>]. "
        "Answer concisely in plain text. If the history doesn't contain "
        "the answer, say so plainly."
    )


def _http(url: str, headers: dict, payload: dict) -> dict:
    try:
        r = requests.post(url, headers=headers, json=payload,
                          timeout=PROVIDER_TIMEOUT)
    except requests.RequestException as e:
        raise AskError(f"provider unreachable: {e}") from e
    if r.status_code != 200:
        first = (r.text or "").strip().splitlines()
        raise AskError(f"provider error {r.status_code}"
                       + (f": {first[0][:200]}" if first else ""))
    return r.json()


class _OpenAI:
    """Chat Completions with tools — the shape most providers speak."""

    def __init__(self, preset):
        self.preset = preset
        base = (preset.get("base_url") or "https://api.openai.com/v1")
        self.url = base.rstrip("/") + "/chat/completions"

    def start(self, system, question):
        self.messages = [{"role": "system", "content": system},
                         {"role": "user", "content": question}]

    def step(self, tools):
        headers = {}
        if self.preset["_key"]:
            headers["Authorization"] = f"Bearer {self.preset['_key']}"
        data = _http(self.url, headers, {
            "model": self.preset["model"], "messages": self.messages,
            # bound the output like the Anthropic adapter: gateways
            # (OpenRouter) credit-check against the model max otherwise
            "max_tokens": int(self.preset.get("max_tokens", 2048)),
            "tools": [{"type": "function",
                       "function": {"name": t["name"],
                                    "description": t["description"],
                                    "parameters": t["schema"]}}
                      for t in tools]})
        msg = data["choices"][0]["message"]
        self.messages.append(msg)
        calls = []
        for c in msg.get("tool_calls") or []:
            try:
                args = json.loads(c["function"]["arguments"] or "{}")
            except json.JSONDecodeError:
                args = {}
            calls.append({"id": c["id"], "name": c["function"]["name"],
                          "args": args})
        u = data.get("usage") or {}
        return {"text": msg.get("content") or "", "tool_calls": calls,
                "usage": {"in": u.get("prompt_tokens", 0),
                          "out": u.get("completion_tokens", 0)}}

    def add_results(self, results):
        for r in results:
            self.messages.append({"role": "tool", "tool_call_id": r["id"],
                                  "content": r["result"]})


class _Anthropic:
    """The Messages API — different tool-call and result shapes."""

    def __init__(self, preset):
        self.preset = preset
        base = preset.get("base_url") or "https://api.anthropic.com"
        self.url = base.rstrip("/") + "/v1/messages"

    def start(self, system, question):
        self.system = system
        self.messages = [{"role": "user", "content": question}]

    def step(self, tools):
        data = _http(self.url, {"x-api-key": self.preset["_key"],
                                "anthropic-version": "2023-06-01"}, {
            "model": self.preset["model"],
            "max_tokens": int(self.preset.get("max_tokens", 2048)),
            "system": self.system, "messages": self.messages,
            "tools": [{"name": t["name"], "description": t["description"],
                       "input_schema": t["schema"]} for t in tools]})
        self.messages.append({"role": "assistant", "content": data["content"]})
        text, calls = "", []
        for block in data["content"]:
            if block["type"] == "text":
                text += block["text"]
            elif block["type"] == "tool_use":
                calls.append({"id": block["id"], "name": block["name"],
                              "args": block.get("input") or {}})
        u = data.get("usage") or {}
        return {"text": text, "tool_calls": calls,
                "usage": {"in": u.get("input_tokens", 0),
                          "out": u.get("output_tokens", 0)}}

    def add_results(self, results):
        self.messages.append({"role": "user", "content": [
            {"type": "tool_result", "tool_use_id": r["id"],
             "content": r["result"]} for r in results]})


_ADAPTERS = {"openai-compatible": _OpenAI, "anthropic": _Anthropic}


def citations(text: str) -> list:
    seen, out = set(), []
    for cid in CITE_RE.findall(text or ""):
        if cid not in seen:
            seen.add(cid)
            out.append(cid)
    return out


def ask(question: str, preset_name: str = "") -> dict:
    """One question, one answer: {answer, citations, usage} or {error}."""
    ps = presets()
    if not ps:
        return {"error": "ask mode isn't configured — add [[ask.models]] "
                         "presets and set their key env vars"}
    if preset_name:
        preset = next((p for p in ps if p["name"] == preset_name), None)
        if preset is None:
            return {"error": f"unknown model preset {preset_name!r}; "
                    "available: " + ", ".join(p["name"] for p in ps)}
    else:
        preset = ps[0]
    backend = preset.get("backend", "openai-compatible")
    if backend not in _ADAPTERS:
        return {"error": f"unknown backend {backend!r} in preset "
                f"{preset['name']!r}"}
    adapter = _ADAPTERS[backend](preset)
    max_turns = int(config.get("ask", "max_turns",
                               "CLAUDE_RAG_ASK_MAX_TURNS", 8))
    tools = tool_defs()
    adapter.start(_system_prompt(), question)
    usage = {"in": 0, "out": 0, "turns": 0, "model": preset["name"]}
    text = ""
    for _ in range(max_turns):
        usage["turns"] += 1
        try:
            step = adapter.step(tools)
        except AskError as e:
            return {"error": str(e), "usage": usage}
        usage["in"] += step["usage"]["in"]
        usage["out"] += step["usage"]["out"]
        text = step["text"] or text
        if not step["tool_calls"]:
            return {"answer": step["text"], "citations":
                    citations(step["text"]), "usage": usage}
        adapter.add_results(
            [{"id": c["id"], "name": c["name"],
              "result": run_tool(c["name"], c["args"])}
             for c in step["tool_calls"]])
    return {"answer": text, "note": "stopped at max_turns",
            "citations": citations(text), "usage": usage}

File: deploy/convex/convex/ask.ts
/* Ask mode, native: a model works the history tools in-process and
   synthesizes an answer — the same loop as ask.py on the Mac, with the
   tools bound to Convex functions. Provider adapters speak raw fetch:
   "openai-compatible" (OpenRouter, OpenAI, Groq …) and "anthropic".
   Presets come from ASK_MODELS; the client only ever names a preset, so
   no argument can point the loop (and a key) at an arbitrary endpoint.
   Tools are read-only, so injected text in indexed content can at worst
   produce a bad answer. */
import { v } from "convex/values";
import { action, internalAction, type ActionCtx } from "./_generated/server";
import { internal } from "./_generated/api";
import { requireUserAction } from "./auth";
import { runSearch, ALL_SOURCES } from "./search";
import { presets, type Preset } from "./archive";

const TOOL_RESULT_MAX = 20_000;
const MAX_TURNS = 8;
const CITE_RE = /\[id:([^\]\s]+)\]/g;

type Tool = { name: string; description: string; schema: Record<string, unknown> };
const TOOLS: Tool[] = [
  {
    name: "search_history",
    description: "Semantic search over the user's own history. Sources: " + ALL_SOURCES.join(", ") +
      ". Returns ranked chunks with id, source, day, text; cite ids as [id:<id>]. " +
      "Use since/until (local YYYY-MM-DD) to window; 'digest' chunks summarise a day per stream.",
    schema: { type: "object", properties: {
      query: { type: "string" }, k: { type: "integer", default: 8 },
      source: { type: "string", description: "restrict to one source" },
      since: { type: "string" }, until: { type: "string" } }, required: ["query"] },
  },
  {
    name: "list_window",
    description: "Everything in a time window, newest first, no ranking — for 'what happened on <day>' " +
      "questions. summaries=true returns only the day digests. At least one bound is required.",
    schema: { type: "object", properties: {
      since: { type: "string" }, until: { type: "string" }, source: { type: "string" },
      summaries: { type: "boolean" }, limit: { type: "integer", default: 50 } } },
  },
  {
    name: "expand",
    description: "Read one chunk in full with its surroundings (conversation turns, the whole note, " +
      "the day's agenda …). Use before leaning on a search hit.",
    schema: { type: "object", properties: { id: { type: "string" }, context: { type: "integer", default: 5 } },
      required: ["id"] },
  },
];

function systemPrompt(strict: boolean) {
  const now = new Date();
  let p = "You answer questions from the user's own indexed history — their Claude Code sessions, " +
    "shell commands, browsing, git commits, notes, tasks, email, calendar, and app usage — using " +
    `the provided tools.\nNow: ${now.toISOString()} (UTC; the user's days are local). Resolve relative dates against that.\n` +
    "Work the disclosure ladder: search_history / list_window to find (read digest chunks first " +
    "for day/week questions); expand to read a hit in full before leaning on it.\n" +
    "Cite the chunks your answer rests on inline as [id:<chunk id>]. Answer concisely in plain " +
    "text. If the history doesn't contain the answer, say so plainly.";
  if (strict) p += "\nDo not answer from general knowledge: search the history first, and every claim " +
    "must cite a chunk it rests on as [id:<chunk id>]. If the history has nothing relevant, say " +
    "exactly that instead of answering.";
  return p;
}

type Call = { id: string; name: string; args: Record<string, unknown> };
type Step = { text: string; calls: Call[]; usage: { in: number; out: number } };

class AskError extends Error {}

async function http(url: string, headers: Record<string, string>, payload: unknown): Promise<any> {
  let r: Response;
  try {
    r = await fetch(url, { method: "POST", headers: { "Content-Type": "application/json", ...headers },
                           body: JSON.stringify(payload) });
  } catch (e) { throw new AskError(`provider unreachable: ${e}`); }
  if (r.status !== 200) {
    const first = (await r.text()).trim().split("\n")[0] ?? "";
    throw new AskError(`provider error ${r.status}${first ? `: ${first.slice(0, 200)}` : ""}`);
  }
  return r.json();
}

interface Adapter {
  start(system: string, question: string): void;
  step(): Promise<Step>;
  addResults(results: { id: string; result: string }[]): void;
}

class OpenAI implements Adapter {
  private messages: any[] = [];
  private url: string;
  constructor(private preset: Preset & { _key: string }, private tools: Tool[] = TOOLS) {
    this.url = (preset.base_url ?? "https://api.openai.com/v1").replace(/\/$/, "") + "/chat/completions";
  }
  start(system: string, question: string) {
    this.messages = [{ role: "system", content: system }, { role: "user", content: question }];
  }
  async step(): Promise<Step> {
    const headers: Record<string, string> = this.preset._key ? { Authorization: `Bearer ${this.preset._key}` } : {};
    const data = await http(this.url, headers, {
      model: this.preset.model, messages: this.messages,
      max_tokens: Number(this.preset.max_tokens ?? 2048),
      ...(this.tools.length ? { tools: this.tools.map((t) =>
        ({ type: "function", function: { name: t.name, description: t.description, parameters: t.schema } })) } : {}),
    });
    const msg = data.choices[0].message;
    this.messages.push(msg);
    const calls: Call[] = [];
    for (const c of msg.tool_calls ?? []) {
      let args = {};
      try { args = JSON.parse(c.function.arguments || "{}"); } catch { /* malformed → no args */ }
      calls.push({ id: c.id, name: c.function.name, args });
    }
    const u = data.usage ?? {};
    return { text: msg.content ?? "", calls, usage: { in: u.prompt_tokens ?? 0, out: u.completion_tokens ?? 0 } };
  }
  addResults(results: { id: string; result: string }[]) {
    for (const r of results) this.messages.push({ role: "tool", tool_call_id: r.id, content: r.result });
  }
}

class Anthropic implements Adapter {
  private messages: any[] = [];
  private system = "";
  private url: string;
  constructor(private preset: Preset & { _key: string }, private tools: Tool[] = TOOLS) {
    this.url = (preset.base_url ?? "https://api.anthropic.com").replace(/\/$/, "") + "/v1/messages";
  }
  start(system: string, question: string) {
    this.system = system;
    this.messages = [{ role: "user", content: question }];
  }
  async step(): Promise<Step> {
    const data = await http(this.url, { "x-api-key": this.preset._key, "anthropic-version": "2023-06-01" }, {
      model: this.preset.model, max_tokens: Number(this.preset.max_tokens ?? 2048),
      system: this.system, messages: this.messages,
      ...(this.tools.length ? { tools: this.tools.map((t) =>
        ({ name: t.name, description: t.description, input_schema: t.schema })) } : {}),
    });
    this.messages.push({ role: "assistant", content: data.content });
    let text = ""; const calls: Call[] = [];
    for (const b of data.content) {
      if (b.type === "text") text += b.text;
      else if (b.type === "tool_use") calls.push({ id: b.id, name: b.name, args: b.input ?? {} });
    }
    const u = data.usage ?? {};
    return { text, calls, usage: { in: u.input_tokens ?? 0, out: u.output_tokens ?? 0 } };
  }
  addResults(results: { id: string; result: string }[]) {
    this.messages.push({ role: "user", content: results.map((r) => ({ type: "tool_result", tool_use_id: r.id, content: r.result })) });
  }
}

const ADAPTERS: Record<string, new (p: Preset & { _key: string }, tools?: Tool[]) => Adapter> = {
  "openai-compatible": OpenAI, anthropic: Anthropic,
};

/* One completion, no tool loop — the command parser's and bench's path
   (wip/SPEC-llm-actions.md). */
export async function chatOnce(
  preset: Preset & { _key: string }, system: string, user: string, maxTokens = 1000,
): Promise<{ text: string; usage: { in: number; out: number } }> {
  const Adapter = ADAPTERS[preset.backend ?? "openai-compatible"];
  if (!Adapter) throw new AskError(`unknown backend '${preset.backend}' in preset '${preset.name}'`);
  const adapter = new Adapter({ ...preset, max_tokens: maxTokens }, []);
  adapter.start(system, user);
  const step = await adapter.step();
  return { text: step.text, usage: step.usage };
}

/* The parser's preset: the one flagged "parse" (set by the bench's
   winner), else fastest by advertised latency. */
export function parsePreset(): (Preset & { _key: string }) | null {
  const ps = presets();
  if (!ps.length) return null;
  const flagged = ps.find((p) => p.parse);
  if (flagged) return flagged;
  const lat = (m: { latency?: string }) => Number((m.latency ?? "").replace(/\D/g, "")) || 99;
  return [...ps].sort((a, b) => lat(a) - lat(b))[0];
}

export function citations(text: string): string[] {
  const seen = new Set<string>(); const out: string[] = [];
  for (const m of text.matchAll(CITE_RE)) if (!seen.has(m[1])) { seen.add(m[1]); out.push(m[1]); }
  return out;
}

export type AskArgs = { q: string; model?: string; strict?: boolean };
export type AskResult = {
  answer?: string; citations?: string[]; note?: string; error?: string;
  usage?: { in: number; out: number; turns: number; model: string };
};

const askArgs = { q: v.string(), model: v.optional(v.string()), strict: v.optional(v.boolean()) };
export const ask = action({
  args: askArgs,
  handler: async (ctx, a): Promise<AskResult> => { await requireUserAction(ctx); return askCore(ctx, a); },
});
export const askInternal = internalAction({
  args: askArgs,
  handler: async (ctx, a): Promise<AskResult> => askCore(ctx, a),
});

export async function askCore(ctx: ActionCtx, a: AskArgs): Promise<AskResult> {
  {
    const ps = presets();
    if (!ps.length) return { error: "ask mode isn't configured — set ASK_MODELS and the key env vars" };
    const preset = a.model ? ps.find((p) => p.name === a.model) : ps[0];
    if (!preset) return { error: `unknown model preset '${a.model}'; available: ${ps.map((p) => p.name).join(", ")}` };
    const Adapter = ADAPTERS[preset.backend ?? "openai-compatible"];
    if (!Adapter) return { error: `unknown backend '${preset.backend}' in preset '${preset.name}'` };

    const runTool = async (name: string, args: Record<string, any>): Promise<string> => {
      try {
        let result: unknown;
        if (name === "search_history") {
          const r = await runSearch(ctx, { query: String(args.query ?? ""), limit: Number(args.k ?? 8),
            sources: args.source ? [String(args.source)] : undefined,
            since: args.since ? String(args.since) : undefined, until: args.until ? String(args.until) : undefined });
          result = { count: r.results.length, results: r.results.map((x, i) => ({
            rank: i + 1, id: x.id, source: x.source, score: x.score, day: x.day, location: x.location,
            text: x.text, meta: x.meta })) };
        } else if (name === "list_window") {
          result = await ctx.runQuery(internal.archive.windowInternal, {
            since: args.since ? String(args.since) : undefined, until: args.until ? String(args.until) : undefined,
            source: args.source ? String(args.source) : undefined, summaries: !!args.summaries,
            limit: Number(args.limit ?? 50) });
        } else if (name === "expand") {
          result = await ctx.runQuery(internal.archive.expandInternal, { id: String(args.id ?? ""), context: Number(args.context ?? 5) });
        } else {
          result = { error: `unknown tool '${name}'` };
        }
        const s = JSON.stringify(result);
        return s.length > TOOL_RESULT_MAX ? s.slice(0, TOOL_RESULT_MAX) + "… [truncated]" : s;
      } catch (e) {
        return JSON.stringify({ error: `${name} failed: ${e}` });
      }
    };

    const adapter = new Adapter(preset);
    adapter.start(systemPrompt(!!a.strict), a.q);
    const usage = { in: 0, out: 0, turns: 0, model: preset.name };
    let text = "";
    for (let turn = 0; turn < MAX_TURNS; turn++) {
      usage.turns++;
      let step: Step;
      try { step = await adapter.step(); }
      catch (e) { return { error: e instanceof AskError ? e.message : String(e), usage }; }
      usage.in += step.usage.in; usage.out += step.usage.out;
      text = step.text || text;
      if (!step.calls.length) return { answer: step.text, citations: citations(step.text), usage };
      const results = [];
      for (const c of step.calls) results.push({ id: c.id, result: await runTool(c.name, c.args) });
      adapter.addResults(results);
    }
    return { answer: text, note: "stopped at max_turns", citations: citations(text), usage };
  }
}
